Celo Standards Checklist (#4669)
celo-org/celo-monorepo: packages/docs/developer-resources/integrations/checklist.md

# Checklist

This page serves as a checklist for applications building and integrating on Celo. Using common features and standards across applications in the Celo ecosystem will lead to a superior user experience.

## General

### Addresses

> Addresses are similar to Ethereum addresses. When displaying and asking for user-inputted addresses, consider using and validating address checksums following the [EIP55 standard](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-55.md) to detect typos.

Reviewer: ```suggestion
Addresses are identical to Ethereum addresses. When displaying and asking for user-inputted addresses, consider using and validating address checksums following the [EIP55 standard](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-55.md) to detect typos.
```

> Addresses are identical to Ethereum addresses. When displaying and asking for user-inputted addresses, consider using and validating address checksums following the [EIP55 standard](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-55.md) to detect typos.

Reviewer 2: could we add a note here about looking up addresses through the registry rather than hardcoding them?

Author: Added a note!

### QR Codes

Celo has [WIP QR code standard CIP16](https://github.com/celo-org/celo-proposals/pull/40) that aims to standardize how applications can ask wallets for transactions to avoid the user having to manually copy/paste addresses and other transaction metadata.

## Custodian/Exchange

Please read more under [Custody](./custody.md), but here is a shortened version:

### Detect Transfers

> cUSD is a token that can be accessed via the ERC20 interface, however the native asset CELO can be accessed via the `GoldToken` ERC20 interface, or natively, similar to `ETH` on Ethereum.

Reviewer: ```suggestion
Stable-value currencies, currently just cUSD, is a contract `StableToken` that can be accessed via the ERC20 interface. The native asset CELO can be accessed via the `GoldToken` ERC20 interface, or natively, similar to ETH on Ethereum.
```

### Proof of Stake
Users may want to participate in Celo's Proof of Stake system to help secure the network and earn rewards.

### Authorized Signers

> Celo's core smart contracts use Celo's `Accounts` abstraction to allow balance-moving keys to be held in cold storage, while other keys can be authorized to vote and held in warm storage or even with the user.

Reviewer: ```suggestion
Celo's core smart contracts use Celo's `Accounts` abstraction to allow balance-moving keys to be held in cold storage, while other keys can be authorized to vote and be held in warm storage or online.
```

### Release Gold

> There is an audited `ReleaseGold` smart contract which allows for the release of `CELO` over a set schedule through which `CELO` might be distributed to a user.

Reviewer: ```suggestion
There is an audited `ReleaseGold` smart contract which allows for the release of CELO over a set schedule through which CELO might be distributed to a user.
```

## Wallets

> As wallets, we are loosely defining any application that is meant as key store to allow users to interact and transfer value on the Celo platform.

Reviewer: ```suggestion
These suggestions apply to any application that custodies a key and allows users to interact and transfer value on the Celo platform.
```

### Key Derivation
Celo wallets should follow the [BIP44](https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki) for deriving private keys from [BIP39 mnemonics](https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki). Celo's key derivation path is at `m/44'/52752'/0'/0`. The first key typically is the `account key` that wallets should register themselves with and accept balance transfers on. The second key can be derived to be an account's `dataEncryptionKey` to allow other users on Celo to encrypt information to.

### Identity Protocol
Celo has a [lightweight identity protocol](/celo-codebase/protocol/identity) that allows users to address each other via their phone number instead of addressses that Celo wallets should implement. Since user privacy is important, Celo wallets should leverage the built-in [Phone Number Privacy protocol](/celo-codebase/protocol/identity/phone-number-privacy) to protect against large-scale harvesting of user phone numbers.

### Wallet Address

> When transferring assets to an account, wallets should check the receiving account's `walletAddress` at which they want to receive funds at. Use cases might be smart contract accounts that want different recovery characteristics, but receive funds at a different address. Also, `walletAddresses` of `0x0` should indicate that the account requires a different mechanism to acquire the `walletAddress`.

Reviewer: explain `walletAddress`

Author: Is that not explained in the first sentence? How would you change it?

### Transaction metadata

> cUSD (aka StableToken) adds an additional method to the ERC20 interface called `transferWithComment` which allows senders to specify an additional comment that Celo wallets should support. Additionally, comments should be encrypted to the `dataEncryptionKey` when applicable.

Reviewer: ```suggestion
The `StableToken` contract that implements cUSD adds an additional method to the ERC20 interface called `transferWithComment` which allows senders to specify an additional comment that Celo wallets should support. Additionally, comments should be encrypted to the `dataEncryptionKey` when applicable.
```

Reviewer: Is there a draft standard for tx metadata we can point to yet?

Author: Not yet. I hope it would change with CIP8 but since that's not operational yet, I keep it more open ended here

> ### Dappkit

Reviewer: link

The Celo ecosystem relies on a diverse set of applications to be built so that users can interact with them. To ensure interopebility and avoid fragmentation due to proprietary interfaces, wallets should be implementing the deeplink-based dappkit to have a consistent interface for dapp developers to implement.

## Validator Group Explorers

> Validator Group Explorers are critical to Celo's Proof of Stake system. Explorers will consider using the following standards to provide a minimum experience across all explorers

Reviewer: ```suggestion
[Validator Group Explorers](../../celo-holder-guide/voting-validators.md#validator-and-validator-group-explorers) are critical to Celo's Proof of Stake system. Explorers will consider using the following standards to provide a minimum experience across all explorers.
```

### Names

All Celo accounts on `Accounts.sol` can claim any name they want. While explorers should display it, they should also be coignizant of the fraud potential.

### Identities

Celo accounts can make claims to existing identities, some of which are verifiable (Domain Names or Keybase profiles). Explorers should consider displaying those identities to reduce the potential for impersonation.

### Performance indicators

> Validator Groups and their validators can perform their duties differently and explorers should reflect that to allow voters to ensure an optimal validator set. While uptime in the form of block signatures by the validators ultimately affect rewards, explorers should also consider displaying other metrics that impact the success of the Celo ecosystem, such as validators' performance in the identity protocol.

Reviewer: ```suggestion
Validator Groups and their validators can perform their duties differently and explorers should reflect that to allow voters to ensure an optimal validator set. While uptime in the form of block signatures by the validators ultimately affect rewards, explorers should also consider displaying [other metrics](../../celo-holder-guide/voting-validators.md#choosing-a-validator-group) that impact the success of the Celo ecosystem, such as validators' performance in the identity protocol.
```
